zebROS_ws/src/behaviors/src/2023_align_to_grid.cpp

```diff
@@ -0,0 +1,231 @@
+// so theoretically the best way to do this is to use the values in 2023_grid_locations.yaml to drive to the location knowing our field-relative location using an apriltag.
+// that's just a particle filter.
+// which we don't trust yet.
+// so we can do the equivalent by calculating the offset between a detected apriltag (see 2023_apriltag_locations.yaml) and the desired location
+// and averaging if there are multiple.
+// find closest apriltag to desired location, calculate offset, call path_to_apriltag
+#include "ros/ros.h"
+#include "field_obj/Detection.h"
+#include "geometry_msgs/Point.h"
+#include <ros/ros.h>
+#include <actionlib/server/simple_action_server.h>
+#include "behavior_actions/AlignToGrid2023Action.h"
+#include "behavior_actions/PathToAprilTagAction.h"
+#include <actionlib/client/simple_action_client.h>
+#include <tf2/LinearMath/Quaternion.h>
+#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
+
+inline geometry_msgs::Point operator-(const geometry_msgs::Point& lhs, const geometry_msgs::Point& rhs) {
+    geometry_msgs::Point p;
+    p.x = lhs.x - rhs.x;
+    p.y = lhs.y - rhs.y;
+    p.z = lhs.z - rhs.z;
+    return p;
+}
+
+inline geometry_msgs::Point operator+(const geometry_msgs::Point& lhs, const geometry_msgs::Point& rhs) {
+    geometry_msgs::Point p;
+    p.x = lhs.x + rhs.x;
+    p.y = lhs.y + rhs.y;
+    p.z = lhs.z + rhs.z;
+    return p;
+}
+
+double distance(const geometry_msgs::Point &pt) {
+    return sqrt(pt.x * pt.x + pt.y * pt.y + pt.z * pt.z);
+}
+
+inline bool operator<(const geometry_msgs::Point& lhs, const geometry_msgs::Point& rhs) {
+    return distance(lhs) < distance(rhs);
+}
+
+inline bool operator>(const geometry_msgs::Point& lhs, const geometry_msgs::Point& rhs) {
+    return distance(lhs) > distance(rhs);
+}
+
+class Tag {
+public:
+	geometry_msgs::Point location;
+	double rotation;
+	Tag(geometry_msgs::Point location = geometry_msgs::Point(), bool _rotation = 0) {
+		this->location = location;
```

Comment:
this-> is redundant in almost all cases - if you want to make it obvious a var is a member, stick with the pattern in the rest of the code where we add an _ to the end of the variable name